include/bbp/sonata/hdf5_reader.h

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+#include <tuple>
+#include <vector>
+
+#include <bbp/sonata/selection.h>
+#include <highfive/H5File.hpp>
+
+namespace bbp {
+namespace sonata {
+
+template <class T>
+class Hdf5PluginReadInterface
+{
+  public:
+    virtual std::vector<T> readSelection(const HighFive::DataSet& dset,
+                                         const Selection& selection) const = 0;
+};
+
+template <class Tuple>
+class Hdf5PluginInterface;
+
+template <class... Types>
+class Hdf5PluginInterface<std::tuple<Types...>>: virtual public Hdf5PluginReadInterface<Types>...
```

Comment:
While I'm not particularly fond of this way of doing things, it's a lot cleaner IMO than using macros to instantiate all the readers. Cheers!

Reply:
That's my favourite part! It's a virtual template method =)

Reply:
My issue is more with C++ itself. But this uses less C, so I'm happy :smile: